File: front/src/__tests__/agentEventReducer.golden.test.ts
import { describe, expect, it } from "vitest";
import { AgentEventReducer, type AgentUIEvent } from "../conversation/runtime/agentEventReducer";
import type { AssistantTurnData, AssistantTurnItem } from "../conversation/rendering/types";
import fixture01 from "../../../contracts/agent-runtime/v1/fixtures/01_simple_text.json";
import fixture02 from "../../../contracts/agent-runtime/v1/fixtures/02_reasoning_text.json";
import fixture03 from "../../../contracts/agent-runtime/v1/fixtures/03_tool.json";
import fixture04 from "../../../contracts/agent-runtime/v1/fixtures/04_parallel_tools.json";
import fixture05 from "../../../contracts/agent-runtime/v1/fixtures/05_tool_failure.json";
import fixture06 from "../../../contracts/agent-runtime/v1/fixtures/06_interrupt.json";
import fixture07 from "../../../contracts/agent-runtime/v1/fixtures/07_reconnect.json";
import fixture08 from "../../../contracts/agent-runtime/v1/fixtures/08_approval.json";
import fixture09 from "../../../contracts/agent-runtime/v1/fixtures/09_retry.json";
import fixture10 from "../../../contracts/agent-runtime/v1/fixtures/10_history_replay.json";
import fixture11 from "../../../contracts/agent-runtime/v1/fixtures/11_duplicate_event.json";
import fixture12 from "../../../contracts/agent-runtime/v1/fixtures/12_out_of_order_event.json";

type FixtureBlock = {
  blockId: string;
  blockSequence: number;
  type: string;
  status: string;
  content?: string;
  callId?: string;
  arguments?: string;
  result?: string;
};

type FixtureTurn = {
  turnId: string;
  turnSequence: number;
  executionId: string;
  status: string;
  blocks: FixtureBlock[];
};

type Fixture = {
  name: string;
  version: number;
  initialState: { lastEventSequence: number; turns: FixtureTurn[] };
  events: AgentUIEvent[];
  expectedState: { lastEventSequence: number; activeTurnId: string; turns: FixtureTurn[] };
  expectedApplyResults?: string[];
};

const fixtures = [
  fixture01, fixture02, fixture03, fixture04, fixture05, fixture06,
  fixture07, fixture08, fixture09, fixture10, fixture11, fixture12,
] as unknown as Fixture[];

function initialItem(turnId: string, block: FixtureBlock): AssistantTurnItem {
  return {
    id: block.blockId,
    turnId,
    conversationId: "conversation-1",
    sequence: block.blockSequence,
    type: block.type,
    status: block.status,
    content: block.content || "",
    callId: block.callId,
    argumentsJson: block.arguments || "",
    resultJson: block.result || "",
  };
}

function initialTurn(turn: FixtureTurn): AssistantTurnData {
  return {
    id: turn.turnId,
    conversationId: "conversation-1",
    executionId: turn.executionId,
    sequence: turn.turnSequence,
    status: turn.status,
    items: turn.blocks.map((block) => initialItem(turn.turnId, block)),
  };
}

function normalizedBlock(item: AssistantTurnItem): FixtureBlock {
  const result: FixtureBlock = {
    blockId: item.id,
    blockSequence: Number(item.sequence || 0),
    type: String(item.type || ""),
    status: String(item.status || ""),
  };
  if (item.content) result.content = item.content;
  if (item.callId) result.callId = item.callId;
  if (item.argumentsJson) result.arguments = item.argumentsJson;
  if (item.resultJson) result.result = item.resultJson;
  return result;
}

function normalizedTurn(turn: AssistantTurnData): FixtureTurn {
  return {
    turnId: turn.id,
    turnSequence: Number(turn.sequence || 0),
    executionId: String(turn.executionId || ""),
    status: String(turn.status || ""),
    blocks: (turn.items || []).map(normalizedBlock),
  };
}

describe("Agent Runtime v1 golden fixtures", () => {
  for (const fixture of fixtures) {
    it(fixture.name, () => {
      expect(fixture.version).toBe(1);
      const reducer = new AgentEventReducer();
      reducer.reset(
        fixture.initialState.turns.map(initialTurn),
        fixture.initialState.lastEventSequence,
        null,
      );
      const results = fixture.events.map((event) => reducer.apply(event));
      if (fixture.expectedApplyResults) expect(results).toEqual(fixture.expectedApplyResults);
      expect({
        lastEventSequence: reducer.lastEventSequence,
        activeTurnId: reducer.activeTurnId,
        turns: reducer.turns.map(normalizedTurn),
      }).toEqual(fixture.expectedState);
    });
  }
});

File: mobile_app/test/features/chat/agent_event_reducer_golden_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:amitia_app/core/models/conversation.dart';
import 'package:amitia_app/features/chat/runtime/agent_event_reducer.dart';
import 'package:flutter_test/flutter_test.dart';

Directory _fixtureDirectory() {
  final candidates = <Directory>[
    Directory('../contracts/agent-runtime/v1/fixtures'),
    Directory('contracts/agent-runtime/v1/fixtures'),
    Directory('../../contracts/agent-runtime/v1/fixtures'),
  ];
  return candidates.firstWhere(
    (candidate) => candidate.existsSync(),
    orElse: () => throw StateError('Agent Runtime v1 golden fixtures not found'),
  );
}

AssistantTurnItemDto _initialItem(String turnId, Map<String, dynamic> block) {
  final type = (block['type'] ?? '').toString();
  return AssistantTurnItemDto(
    id: (block['blockId'] ?? '').toString(),
    turnId: turnId,
    conversationId: 'conversation-1',
    sequence: (block['blockSequence'] as num?)?.toInt() ?? 0,
    type: type,
    status: (block['status'] ?? '').toString(),
    content: (block['content'] ?? '').toString(),
    callId: (block['callId'] ?? '').toString(),
    argumentsJson: (block['arguments'] ?? '').toString(),
    resultJson: (block['result'] ?? '').toString(),
  );
}

AssistantTurnDto _initialTurn(Map<String, dynamic> turn) {
  final turnId = (turn['turnId'] ?? '').toString();
  final rawBlocks = turn['blocks'];
  return AssistantTurnDto(
    id: turnId,
    conversationId: 'conversation-1',
    executionId: (turn['executionId'] ?? '').toString(),
    sequence: (turn['turnSequence'] as num?)?.toInt() ?? 0,
    status: (turn['status'] ?? '').toString(),
    items: rawBlocks is List
        ? rawBlocks
            .whereType<Map>()
            .map((raw) => _initialItem(turnId, Map<String, dynamic>.from(raw)))
            .toList(growable: false)
        : const <AssistantTurnItemDto>[],
  );
}

Map<String, dynamic> _normalizedBlock(AssistantTurnItemDto item) {
  final value = <String, dynamic>{
    'blockId': item.id,
    'blockSequence': item.sequence,
    'type': item.type,
    'status': item.status,
  };
  if (item.content.isNotEmpty) value['content'] = item.content;
  if (item.callId.isNotEmpty) value['callId'] = item.callId;
  if (item.argumentsJson.isNotEmpty) value['arguments'] = item.argumentsJson;
  if (item.resultJson.isNotEmpty) value['result'] = item.resultJson;
  return value;
}

Map<String, dynamic> _normalizedTurn(AssistantTurnDto turn) => <String, dynamic>{
  'turnId': turn.id,
  'turnSequence': turn.sequence,
  'executionId': turn.executionId,
  'status': turn.status,
  'blocks': turn.items.map(_normalizedBlock).toList(growable: false),
};

void main() {
  final files = _fixtureDirectory()
      .listSync()
      .whereType<File>()
      .where((file) => file.path.endsWith('.json'))
      .toList()
    ..sort((left, right) => left.path.compareTo(right.path));

  for (final file in files) {
    final fixture = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
    test('Agent Runtime v1 golden: ${fixture['name']}', () {
      expect(fixture['version'], 1);
      final initial = Map<String, dynamic>.from(fixture['initialState'] as Map);
      final rawInitialTurns = initial['turns'];
      final reducer = AgentEventReducer();
      reducer.reset(
        turns: rawInitialTurns is List
            ? rawInitialTurns
                .whereType<Map>()
                .map((raw) => _initialTurn(Map<String, dynamic>.from(raw)))
                .toList(growable: false)
            : const <AssistantTurnDto>[],
        lastEventSequence: (initial['lastEventSequence'] as num?)?.toInt() ?? 0,
      );

      final results = <String>[];
      final rawEvents = fixture['events'] as List;
      for (final raw in rawEvents.whereType<Map>()) {
        results.add(
          reducer
              .apply(AgentUIEvent.fromJson(Map<String, dynamic>.from(raw)))
              .name,
        );
      }
      final expectedResults = fixture['expectedApplyResults'];
      if (expectedResults is List) {
        expect(results, expectedResults.map((value) => value.toString()).toList());
      }

      expect(
        <String, dynamic>{
          'lastEventSequence': reducer.lastEventSequence,
          'activeTurnId': reducer.activeTurnId,
          'turns': reducer.turns.map(_normalizedTurn).toList(growable: false),
        },
        Map<String, dynamic>.from(fixture['expectedState'] as Map),
      );
    });
  }
}
